arraycontext/container/traversal.py

```diff
@@ -262,12 +263,15 @@ def keyed_map_array_container(f: Callable[[Any, Any], Any],
     :param ary: a (potentially nested) structure of :class:`ArrayContainer`\ s,
         or an instance of a base array type.
     """
-    if is_array_container(ary):
-        return deserialize_container(ary, [
-                (key, f(key, subary)) for key, subary in serialize_container(ary)
-                ])
+    try:
+        iterable = serialize_container(ary)
+    except NotImplementedError:
+        raise ValueError(
+                f"Non-array container type has no key: {type(ary).__name__}")
```

Comment:
I think we may want to distinguish between `ValueError` and `NotImplementedError`. Also, I'm not sure if we necessarily need to catch exceptions if we do proper error handling further down

Reply:
Not quite sure I follow what the suggestion is. Can you elaborate a bit? This code should be the same as before, it just uses `try .. except` instead of `is_array_container`.

Reply:
@alexfikl I think the discussion in this thread https://github.com/inducer/arraycontext/pull/90#discussion_r714292397 settles this!

Reply:
Got it! :grin: 

Thanks a lot for taking a look and sparking a discussion! Always appreciated!